### PR TITLE
Additional WS support

### PR DESCRIPTION
**Pull request by boxhock**
## Changes

- Add Forex support to Finage
- Add Crypto WS support to Tiingo
- Enabled WS for Trademade even when `NODE_ENV` isn't `development`

## Steps to Test

1. Run adapters with WS
2. Send requests

## Quality Assurance

- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `<ADAPTER_PACKAGE>/schemas/env.json` and `<ADAPTER_PACKAGE>/README.md`
- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [x] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Clubhouse/Shortcut
- [x] This is related to a maximum of one Clubhouse/Shortcut story or GitHub issue
- [x] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types)

### CODE REVIEW

File: packages/sources/finage/src/adapter.ts [hunk 1]
```diff
@@ -1,141 +1,95 @@
...
   }
-  const getSymbol = (input: AdapterRequest) => {
-    const validator = new Validator(input, customParams, {}, false)
+  const getStockSymbol = (input: AdapterRequest) => {
+    const validator = new Validator(input, endpoints.stock.inputParams, {}, false)
     if (validator.error) return
     return validator.validated.data.base.toUpperCase()
   }
+  const isStock = (input: AdapterRequest): boolean =>
+    endpoints.stock.supportedEndpoints.indexOf(input.data.endpoint) > -1
```
Comment (austinborn) [inline packages/sources/finage/src/adapter.ts:37]:
```suggestion
  const isStock = (input: AdapterRequest): boolean =>
    endpoints.stock.supportedEndpoints.includes(input.data.endpoint)
```

File: packages/sources/finage/src/adapter.ts [hunk 2]
```diff
@@ -1,141 +1,95 @@
...
+          }
+        }
+        return undefined
       },
       subscribe: (input) => getSubscription(getSymbol(input)),
       unsubscribe: (input) => getSubscription(getSymbol(input), false),
       subsFromMessage: (message) => {
-        if (!message.s) return undefined
-        return getSubscription(`${message.s.toUpperCase()}`)
+        if (message.s) return getSubscription(`${message.s.toUpperCase()}`)
```
Comment (austinborn) [inline packages/sources/finage/src/adapter.ts:79]:
```suggestion
        if (message.s) return getSubscription(message.s.toUpperCase())
```

File: packages/sources/tiingo/src/adapter.ts [hunk 3]
```diff
@@ -75,20 +75,37 @@ export const makeWSHandler = (config?: Config): MakeWSHandler | undefined => {
...
-  const getTicker = (input: AdapterRequest): string | undefined => {
+  const getFxTicker = (input: AdapterRequest): string | undefined => {
     const validator = new Validator(input, customParams, {}, false)
     if (validator.error) return
-    const base = validator.validated.data.base.toLowerCase()
-    return base
+    return validator.validated.data.base.toLowerCase()
+  }
+  const isFx = (input: AdapterRequest): boolean =>
+    endpoints.iex.supportedEndpoints.indexOf(input.data.endpoint) > -1
```
Comment (austinborn) [inline packages/sources/tiingo/src/adapter.ts:84]:
```suggestion
    endpoints.iex.supportedEndpoints.includes(input.data.endpoint)
```

File: packages/sources/tiingo/src/adapter.ts [hunk 4]
```diff
@@ -75,20 +75,37 @@ export const makeWSHandler = (config?: Config): MakeWSHandler | undefined => {
...
+  const getCryptoTicker = (input: AdapterRequest): string | undefined => {
+    const validator = new Validator(input, endpoints.prices.inputParameters, {}, false)
+    if (validator.error) return
+    return `${validator.validated.data.base.toLowerCase()}${validator.validated.data.quote.toLowerCase()}`
```
Comment (austinborn) [inline packages/sources/tiingo/src/adapter.ts:88]:
```suggestion
    const { base, quote } = validator.validated.data
    return `${base}${quote}`.toLowerCase()
```

File: packages/sources/tiingo/src/adapter.ts [hunk 5]
```diff
@@ -75,20 +75,37 @@ export const makeWSHandler = (config?: Config): MakeWSHandler | undefined => {
...
   }
+  const isCrypto = (input: AdapterRequest): boolean =>
+    endpoints.prices.supportedEndpoints.indexOf(input.data.endpoint) > -1
```
Comment (austinborn) [inline packages/sources/tiingo/src/adapter.ts:91]:
```suggestion
    endpoints.prices.supportedEndpoints.includes(input.data.endpoint)
```

File: packages/sources/finage/src/adapter.ts [hunk 6]
```diff
@@ -1,141 +1,95 @@
...
+
+  const getForexSymbol = (input: AdapterRequest) => {
+    const validator = new Validator(input, endpoints.forex.inputParams, {}, false)
+    if (validator.error) return
+    const from = (validator.overrideSymbol(NAME) as string).toUpperCase()
+    const to = validator.validated.data.quote.toUpperCase()
+    return `${from}/${to}` // Note that this adds the "/", whereas the REST endpoint doesn't use this
+  }
+  const isForex = (input: AdapterRequest): boolean =>
+    endpoints.forex.supportedEndpoints.indexOf(input.data.endpoint) > -1
```
Comment (austinborn) [inline packages/sources/finage/src/adapter.ts:47]:
```suggestion
    endpoints.forex.supportedEndpoints.includes(input.data.endpoint)
```

**Review by austinborn — DISMISSED** [review]

File: packages/sources/finage/src/adapter.ts (hunk 1 above)
Comment (boxhock) [reply packages/sources/finage/src/adapter.ts:37]: I swear I tried this, but no idea why I didn't commit this, thanks :laughing:

**Review by austinborn — DISMISSED** [review]

File: packages/sources/finage/src/config.ts [hunk 7]
```diff
@@ -1,16 +1,34 @@
...
+export type Config = BaseConfig & {
+  stockWsEndpoint: string
+  forexWsEndpoint: string
+}
 
 export const makeConfig = (prefix?: string): Config => {
   const config = Requester.getDefaultConfig(prefix, true)
-  config.api.baseURL = config.api.baseURL || DEFAULT_BASE_URL
-  const wsUrl = config.api.baseWsURL || DEFAULT_WS_API_ENDPOINT
+  const stockWsUrl = util.getEnv(ENV_STOCK_WS_API_ENDPOINT) || DEFAULT_STOCK_WS_API_ENDPOINT
```
Comment (justinkaseman) [inline packages/sources/finage/src/config.ts:21]: We don't have a precedence for this yet, but we should document these environment variables in the README. Maybe a new WS section.

(also add to the schema)

**Review by justinkaseman — APPROVED** [review]
Nice. LGTM.

**Review by justinkaseman — APPROVED** [review]

**Pull request merged**